test/boost/aws_error_injection_test.cc

```diff
@@ -0,0 +1,280 @@
+
+    semaphore mem{memory_size};
+    auto client = s3::client::make(address, make_minio_config(port), mem);
+    auto client_shutdown = deferred_close(*client);
+    client->upload_file(file_path, object_name).get();
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multipart_upload_file_success) {
+    auto port = get_port();
+    auto address = get_address();
+    server server(server::failure_policy::SUCCESS);
+    server.start(address, port).get();
+    auto close_server = deferred_stop(server);
+    const size_t part_size = 5_MiB;
+    const size_t remainder_size = part_size / 2;
+    const size_t total_size = 4 * part_size + remainder_size;
+    const size_t memory_size = part_size;
+    BOOST_REQUIRE_NO_THROW(test_client_upload_file(seastar_test::get_name(), address, port, total_size, memory_size));
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multipart_upload_file_retryable_success) {
+    auto port = get_port();
+    auto address = get_address();
+    server server(server::failure_policy::RETRYABLE_FAILURE);
+    server.start(address, port).get();
+    auto close_server = deferred_stop(server);
+    const size_t part_size = 5_MiB;
+    const size_t remainder_size = part_size / 2;
+    const size_t total_size = 4 * part_size + remainder_size;
+    const size_t memory_size = part_size;
+    BOOST_REQUIRE_NO_THROW(test_client_upload_file(seastar_test::get_name(), address, port, total_size, memory_size));
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multipart_upload_file_failure) {
+    auto port = get_port();
+    auto address = get_address();
+    server server(server::failure_policy::NONRETRYABLE_FAILURE);
+    server.start(address, port).get();
+    auto close_server = deferred_stop(server);
+    const size_t part_size = 5_MiB;
+    const size_t remainder_size = part_size / 2;
+    const size_t total_size = 4 * part_size + remainder_size;
+    const size_t memory_size = part_size;
+    BOOST_REQUIRE_EXCEPTION(test_client_upload_file(seastar_test::get_name(), address, port, total_size, memory_size),
+                            storage_io_error,
+                            [](const storage_io_error& e) { return e.code().value() == EIO; });
+}
+
+void do_test_client_multipart_upload(const std::string& address, uint16_t port) {
+    const sstring name(fmt::format("/{}/testobject-{}", "test", "large", ::getpid()));
+
+    testlog.info("Make client");
+    semaphore mem(16 << 20);
+    auto cln = s3::client::make(address, make_minio_config(port), mem);
+    auto close_client = deferred_close(*cln);
+
+    testlog.info("Upload object");
+    auto out = output_stream<char>(cln->make_upload_sink(name));
+
+    static constexpr unsigned chunk_size = 1000;
+    auto rnd = tests::random::get_bytes(chunk_size);
+    for (unsigned ch = 0; ch < 128_KiB; ch++) {
+        out.write(reinterpret_cast<char*>(rnd.begin()), rnd.size()).get();
+    }
+
+    testlog.info("Flush multipart upload");
+    out.flush().get();
+    out.close().get();
```

Comment:
might want to to use `deferred_close()`  as well.

Reply:
Cant. As I understand, once `flush` throws the `close` will fail since we have unconsumed data in the stream.

Reply:
It shouldn't, if flush throws, then there is (shouldn't be) no unsent data left in it

Reply:
Addressed [here](https://github.com/scylladb/scylladb/pull/20518/files#diff-28c1d03a5facd99f25f466e6bae94c0fab2bb8845c0538dd489db7832505e884R785)